web/packages/build/package.json

```diff
@@ -53,6 +53,8 @@
     "@types/react-dom": "^18.2.17",
     "@types/react-router-dom": "^5.1.1",
     "@types/react-transition-group": "^4.4.5",
+    "@types/styled-components": "^5.1.34",
+    "@types/styled-system": "^5.1.22",
```

Comment:
What's the compatibility between those packages and styled-components and styled-system? We're on 5.3.11 of styled-components, so I guess the types are valid for that version of that package. But I'm not sure about styled-system. We use 3.2.1 of styled-system, but this package here is on 5.1.22.

Reply:
I don't think that's the way to proceed with this. External type definitions, such as those in the DefinitelyTyped repo, unlike type definitions directly in .ts files, can tell lies. As in, they are nothing more than statements saying "function `foo` returns `boolean`". TypeScript doesn't verify in any way if they're correct.

That's why it's so important that the definitions match the library version for which they were written for. styled-system went through [many changes between 3.2.1 and 5.1.5](https://github.com/styled-system/styled-system/blob/master/CHANGELOG.md), so I'd expect that the types in the newest package do not work with 3.2.1.

E.g., I see https://github.com/DefinitelyTyped/DefinitelyTyped/commit/5746d0876fca8f306d2319c0bf821511d550300b which updates the compatibility of those definitions to 4.1. The date of this commit corresponds to the release of 4.1.1 of `@types/styled-system` (`npm view @types/styled-system --json`).

I don't want us to run into problems down the line that stem from type definitions being incompatible with the library itself, because it'll be tricky to undo that. That's why I'd like to see a confirmation that 5.1.22 of `@types/styled-system` is indeed compatible with `@styled-system` 3.2.1. If there isn't any, the I think we should stick with the version of the definitions which was made with our version in mind.

---

When I change `@types/styled-system` to 3.2.2, which seems to correspond to our version of `styled-system`, then I do get around 25 extra errors. E.g., it complains about `<Flex border={1}>` being wrong here:

https://github.com/gravitational/teleport/blob/79b0385aa134b4e66a8821a0ab0e3d40cf1864ae/web/packages/shared/components/AccessRequests/NewRequest/RequestCheckout/SelectReviewers.tsx#L190-L193

And that seems to stem from the fact that in `@types/styled-system@3.2.2`, there's `BorderProps` which corresponds to the `border` prop. `Box`, which `Flex` inherits from `BordersProps`, `BorderRadiusProps`, `BorderColorProps`, but not `BorderProps`. Adding it there fixes the problem.

3.2.2 complains about `fontWeight="regular"`, since this is our custom defined font weight, but 3.2.2 has no idea about those. 5.1.22 appears to be fine with it and indeed, [the definitions seem to consider theme values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d8da2b742fb2d05ad786d8afb42d1032c0fa2eff/types/styled-system/index.d.ts#L369-L378). But it turns out it's fine with it only because it accepts any string under that prop – `fontWeight="foo"` works just as fine. `FontWeightProps` accepts some extra type, but we don't provide any when using that type.

https://github.com/gravitational/teleport/blob/acb7510479d76a43ac207bbb435588321b98686f/web/packages/design/src/Text/Text.tsx#L39-L45

One could argue that it's at least not worse than it is right now, but then what do we gain from merging this PR? From what I see it's mostly better handling of custom-defined props on styled components such as in the example below. But are there any other mistakes that those types can catch?

https://github.com/gravitational/teleport/blob/acb7510479d76a43ac207bbb435588321b98686f/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx#L95-L107